core/src/main/java/org/apache/iceberg/ReachableFileUtil.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ReachableFileUtil {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ReachableFileUtil.class);
+
+  private ReachableFileUtil() {
+  }
+
+  /**
+   * Returns the location of the version.text file
+   *
+   * @param ops tableOperation for which version.text path needs to be retrieved
+   * @return the location of the version hint file
+   */
+  public static String versionHintLocation(TableOperations ops) {
+    return ops.metadataFileLocation("version-hint.text");
+  }
+
+  /**
+   * Returns locations of JSON metadata files in a table.
+   *
+   * @param ops       TableOperations to get JSON metadata files from
+   * @param recursive When true, recursively retrieves all the reachable JSON metadata files.
+   *                 When false, gets the all the JSON metadata files only from the current metadata.
+   * @return locations of JSON metadata files
+   */
+  public static Set<String> metadataFileLocations(TableOperations ops, boolean recursive) {
+    Set<String> metadataFileLocations = Sets.newHashSet();
+    TableMetadata tableMetadata = ops.current();
+    metadataFileLocations.add(tableMetadata.metadataFileLocation());
+    metadataFileLocations(tableMetadata, metadataFileLocations, ops.io(), recursive);
+    return metadataFileLocations;
+  }
+
+  private static void metadataFileLocations(TableMetadata metadata, Set<String> metaFiles,
+                                            FileIO io, boolean isRecursive) {
+    List<TableMetadata.MetadataLogEntry> metadataLogEntries = metadata.previousFiles();
+    if (metadataLogEntries.size() > 0) {
+      for (TableMetadata.MetadataLogEntry metadataLogEntry : metadataLogEntries) {
+        metaFiles.add(metadataLogEntry.file());
+      }
+      if (isRecursive) {
```

Comment:
Per off-line discussion, we are trying to find the first existing metadata.json file while recursive searching. Let's take the test as an example,  there are v1 v2 v3 and v4, METADATA_PREVIOUS_VERSIONS_MAX is 2. Inside the v4, you have v2 and v3, and v2 is the first item, if v2 is there, open v2 and continue searching, but if v2.metadata.json is deleted, we can open the second item(v3.metadata.json) and continue searching. It's simple logic-wise and perf is also good since we don't open the same file multiple times.

Reply:
I was doing doing that initially. Based on https://github.com/apache/iceberg/pull/2564#discussion_r630695415, I am now checking the only the oldest file and bailing out if not existent.

Reply:
Thanks @karuppayya. I think your initial way makes sense, for implementation, I'd suggest to wrap the logic into a small method like `findFirstExistedJson`. Let's hear thoughts from @aokolnychyi.

Reply:
I am a bit worried about the number of requests we can send that way. I feel like if the lineage broken then it is fine to give up. That being said, I am also fine if both of you think it is going to be helpful. I agree a separate method should simplify it a bit.

Reply:
Leave this up to you, @karuppayya @flyrain.

Reply:
Add a additional method to find the first existent metadata and using it figure the metadata file locations.

Reply:
Sounds good to me. Resolving this.